apps/frontend/app/(website)/studio/main/ASTViewer/utils.ts

```diff
@@ -0,0 +1,90 @@
+export function extractComments(node: Node): [Node, Node[]] {
+  const comments: Node[] = [];
+  const seenComments = new Set<string>();
+
+  function traverse(node: Node, parent: Node | null = null) {
+    if (typeof node !== "object" || node === null) {
+      return node;
+    }
+
+    if (node.label === "CommentLine" || node.label === "CommentBlock") {
+      const commentKey = `${node.start}-${node.end}`;
+      if (!seenComments.has(commentKey)) {
+        seenComments.add(commentKey);
+        comments.push({
+          ...node,
+          parentId: parent?.id,
+          parentStart: parent?.start,
+          parentEnd: parent?.end,
+        });
+      }
+      return undefined;
+    }
+
+    const result = {};
+    for (const [key, value] of Object.entries(node)) {
+      if (key !== "comments" && key !== "internalComments") {
+        if (Array.isArray(value)) {
+          result[key] = value
+            .map((item) => traverse(item, node))
+            .filter((item) => item !== undefined);
+        } else if (typeof value === "object" && value !== null) {
+          const traversed = traverse(value, node);
+          if (traversed !== undefined) {
+            result[key] = traversed;
+          }
+        } else {
+          result[key] = value;
+        }
+      }
+    }
```

Comment:
reduce with early returns would help avoiding 3 conditional statements and one could argue it's easier to read that way. if you agree, let's switch